src/ROOTDataSource.cc

```diff
@@ -0,0 +1,308 @@
+#include "podio/ROOTDataSource.h"
+
+// STL
+#include <cstddef>
+#include <cstdio>
+#include <exception>
+#include <filesystem>
+#include <iostream>
+#include <memory>
+
+// ROOT
+#include <TFile.h>
+
+// podio
+#include <podio/Frame.h>
+#include <podio/ROOTReader.h>
+
+namespace podio {
+ROOTDataSource::ROOTDataSource(const std::string& filePath, int nEvents) : m_nSlots{1} {
+  m_filePathList.emplace_back(filePath);
+  SetupInput(nEvents);
+}
+
+ROOTDataSource::ROOTDataSource(const std::vector<std::string>& filePathList, int nEvents) :
+    m_nSlots{1}, m_filePathList{filePathList} {
+  SetupInput(nEvents);
+}
+
+/// @TODO Check for the existence of the file, which might be coming from web
+///       or EOS.
+void ROOTDataSource::SetupInput(int nEvents) {
+  // std::cout << "podio::ROOTDataSource: Constructing the source ..." << std::endl;
+
+  if (m_filePathList.empty()) {
+    throw std::runtime_error("podio::ROOTDataSource: No input files provided!");
+  }
+
+  for (const auto& filePath : m_filePathList) {
+    // Check if file exists
+    // if (!std::filesystem::exists(filePath)) {
+    //   throw std::runtime_error("podio::ROOTDataSource: Provided file \""
+    //                            + filePath + "\" does not exist!");
+    // }
+
+    // Check if the provided file contains required metadata
+    std::unique_ptr<TFile> inFile(TFile::Open(filePath.data(), "READ"));
+    auto metadata = inFile->Get("podio_metadata");
+    if (!metadata) {
+      throw std::runtime_error("podio::ROOTDataSource: Provided file is missing podio metadata!");
+    }
+  }
+
+  // Create probing frame
+  podio::Frame frame;
+  unsigned int nEventsInFiles = 0;
+  podio::ROOTReader podioReader;
+  podioReader.openFiles(m_filePathList);
```

Comment:
```suggestion
  auto podioReader = podio::makeReader(m_filePathList);
```
with the reader interface.

Reply:
Done